Core/Src/DynamicControls/longitudinal_control.c

```diff
@@ -0,0 +1,40 @@
+/*
+ * longitudinal_control.c
+ *
+ *  Created on: Sep 15th 2021
+ *      Author: GiovanniPasa
+ */
+
+#include "DynamicControls/longitudinal_control.h"
+#include "DynamicControls/initializer_controls.h"
+#include "DynamicControls/PID.h"
+#include "cmsis_os.h"
+#include "math.h"
+#include "speed_calc.h"
+#include "global_variables.h"
+#include "constants.h"
+
+extern PID_t pid_longitudinal_L;
+extern PID_t pid_longitudinal_R;
+
+longitudinal_t longitudinal_control(volatile float g_wheel_speed[4]) {
+    float cm_speed;
+    double left_slip, right_slip, setpoint;
+    longitudinal_t ref_torque;
+
+    // speed and slip ratios
+    cm_speed = ((g_wheel_speed[FRONT_RIGHT] + g_wheel_speed[FRONT_LEFT])/2);  // speed of the car's center of mass
+    left_slip = ((g_wheel_speed[REAR_LEFT] - cm_speed) / cm_speed) * 100;   // slip ratio of the rear left wheel
+    right_slip = ((g_wheel_speed[REAR_RIGHT] - cm_speed) / cm_speed) * 100;  // slip ratio of the rear right wheel
```

Comment:
eu acho que as rodas traseiras não vão ter mais sensores de roda fônica, talvez vc precise substituir isso pela velocidade do motor.
@FelTell confirma isso pra mim por favor?

Reply:
Não vamos usar a roda fônica. A ideia é usar a velocidade do motor msm, tá no backlog pra ser implementado a leitura e cálculo.
Mas acho que poderia usar essa msm variável n?

Reply:
é porque esse `g_wheel_speed` (q vai virar `WHEEL_SPeED` com aquele rolê das queues) é setado lá no speed_calc, q calcula a velocidade a partir das interrupções do sensor indutivo q fica na roda fônica. Como n vai ter roda fônica atrás, n vai ter input do sensor, aí esse valor pra rodas traseiras vai ficar sempre zerado

Reply:
Lembro que fiz aquele código pensando em que posso limitar pra ele fazer verificação nas duas rodas dianteiras apenas.
Então tem duas opções: usar essa variável mesmo ou modificar para ser INDUCTIVE_WHEEL_SPEED e criar a MOTOR_WHEEL_SPEED (algo assim, tem que diferenciar sensor se for usar duas variáveis)

Reply:
Acho q n dá pra usar a mesma variável pq essa wheel_speed só é setada onde calcula velocidade da roda fônica, e como as de trás vêm do motor, ia precisar estar ela em outro lugar (em quesito de arquitetura do código, é feio isso). Mas já existe a variável MOTOR_SPEED, só q ela nunca é setada. Isso é pq vem da CAN do inversor esse dado, e a gnt nunca tratou os dados recebidos do inversor até hj

Reply:
inclusive é uma coisa q vcs vão precisar colocar no backlog aí, q é tratar o recebimento de dados pela CAN. Eu já tinha preparado a estrutura do código pra receber esses dados e o Telles deu uma mexida nisso tbm, mas dps precisa terminar de implementar pra guardar esses dados, populando as variáveis globais e fazendo datalog

Reply:
> Mas já existe a variável MOTOR_SPEED, só q ela nunca é setada. Isso é pq vem da CAN do inversor esse dado, e a gnt nunca tratou os dados recebidos do inversor até hj

Acho válido mudar o nome pq no fim todas vão ser velocidades da roda.

> Eu já tinha preparado a estrutura do código pra receber esses dados

Tá preparado rs? Temos que conversar depois pra saber qual o melhor jeito de fazer isso. Acho que dá pra fazer algo melhor que um switch case com todas as variáveis né

Reply:
adicionei um TODO pra ser realizado assim q essa parte de recebimento via can/inversor do codigo estiver pronta